x-pack/plugins/observability_solution/apm_data_access/server/utils/unflatten_known_fields.ts

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { DedotObject } from '@kbn/utility-types';
+import * as APM_EVENT_FIELDS_MAP from '@kbn/apm-types/es_fields';
+import type { ValuesType } from 'utility-types';
+import { unflattenObject } from '@kbn/observability-utils/object/unflatten_object';
+import { mergePlainObjects } from '@kbn/observability-utils/object/merge_plain_objects';
+import { castArray, isArray } from 'lodash';
+import { AgentName } from '@kbn/elastic-agent-utils';
+import { EventOutcome } from '@kbn/apm-types/src/es_schemas/raw/fields';
+import { ProcessorEvent } from '@kbn/observability-plugin/common';
+
+const {
+  CLOUD,
+  AGENT,
+  SERVICE,
+  ERROR_EXCEPTION,
+  SPAN_LINKS,
+  HOST,
+  KUBERNETES,
+  CONTAINER,
+  TIER,
+  INDEX,
+  DATA_STEAM_TYPE,
+  VALUE_OTEL_JVM_PROCESS_MEMORY_HEAP,
+  VALUE_OTEL_JVM_PROCESS_MEMORY_NON_HEAP,
+  SPAN_LINKS_SPAN_ID,
+  SPAN_LINKS_TRACE_ID,
+  ...CONCRETE_FIELDS
+} = APM_EVENT_FIELDS_MAP;
+
+const ALL_FIELDS = Object.values(CONCRETE_FIELDS);
+
+const KNOWN_MULTI_VALUED_FIELDS = [
+  APM_EVENT_FIELDS_MAP.CHILD_ID,
+  APM_EVENT_FIELDS_MAP.PROCESS_ARGS,
+] as const;
+
+type KnownField = ValuesType<typeof CONCRETE_FIELDS>;
+
+type KnownSingleValuedField = Exclude<KnownField, KnownMultiValuedField>;
+type KnownMultiValuedField = ValuesType<typeof KNOWN_MULTI_VALUED_FIELDS>;
+
+const KNOWN_SINGLE_VALUED_FIELDS = ALL_FIELDS.filter(
+  (field): field is KnownSingleValuedField => !KNOWN_MULTI_VALUED_FIELDS.includes(field as any)
+);
+
+interface TypeOverrideMap {
+  [APM_EVENT_FIELDS_MAP.SPAN_DURATION]: number;
+  [APM_EVENT_FIELDS_MAP.AGENT_NAME]: AgentName;
```

Comment:
As mentioned in the call, I think that type here might become an issue with the way agent names are stored since recently. See https://github.com/elastic/kibana/issues/180444 for more context

Reply:
good callout. I'll look at it.

Reply:
I think there is a way in typescript to specify that it is a string that can start with the values suggested by AgentNames, but it could contain a prefix, so it's not an entire match

Reply:
@rmyz good idea!

Reply:
That still sounds a bit brittle. What happens when the string doesn't comply with the expected values? Will this cause the entire page to crash? Is there a downside to just allowing all strings here?

Reply:
I think it's safe to type it like this. It will be pretty generic after the `/`,

```ts
export type OpenTelemetryAgentName =
  | 'opentelemetry'
  | 'otlp'
  | `opentelemetry/${string}`
  | `otlp/${string}`;
```

Reply:
Let's look into this in a separate PR. Creating a type with string literals causes problems in the telemetry serializer, which isn't prepared to handle such a type. Replacing `AgentName` with `string` will require changes in other places.

Reply:
I will look into it tomorrow. I just want the build to pass first.